fabric-rendering-v1/src/client/java/net/fabricmc/fabric/api/client/rendering/v1/world/WorldRenderEvents.java

```diff
@@ -0,0 +1,256 @@
+	 * and you don't want the block outline render to appear.
+	 *
+	 * <p>This event should NOT be used for general-purpose replacement of
+	 * the default block outline rendering because it will interfere with mod-specific
+	 * renders.  Mods that replace the default block outline for specific blocks
+	 * should instead subscribe to {@link #BEFORE_BLOCK_OUTLINE}.
+	 */
+	public static final Event<AfterBlockOutlineExtraction> AFTER_BLOCK_OUTLINE_EXTRACTION = EventFactory.createArrayBacked(AfterBlockOutlineExtraction.class, callbacks -> (context, hit) -> {
+		for (final AfterBlockOutlineExtraction callback : callbacks) {
+			callback.afterBlockOutlineExtraction(context, hit);
+		}
+	});
+
+	/**
+	 * Called after the extraction phase is complete.
+	 * Use this to extract general custom data needed for rendering.
+	 *
+	 * <p>To attach modded data to vanilla render states, see {@link net.fabricmc.fabric.api.client.rendering.v1.FabricRenderState FabricRenderState}.
+	 * Only attach the minimum data needed for rendering. Do not attach objects that are not thread-safe such as {@link net.minecraft.client.world.ClientWorld}.
+	 */
+	public static final Event<EndExtraction> END_EXTRACTION = EventFactory.createArrayBacked(EndExtraction.class, callbacks -> context -> {
+		for (final EndExtraction callback : callbacks) {
+			callback.endExtraction(context);
+		}
+	});
+
+	/**
+	 * Called after all chunks to be rendered are uploaded to GPU,
+	 * before any chunks are output to the framebuffer.
+	 */
+	public static final Event<StartMain> START_MAIN = EventFactory.createArrayBacked(StartMain.class, callbacks -> context -> {
+		for (final StartMain callback : callbacks) {
+			callback.startMain(context);
+		}
+	});
+
+	/**
+	 * Called after the Solid, Cutout, and Cutout Mipped terrain layers have been output to the framebuffer.
+	 *
+	 * <p>Use to render non-translucent terrain to the framebuffer.
+	 *
+	 * <p>Note that 3rd-party renderers may combine these passes or otherwise alter the
+	 * rendering pipeline for sake of performance or features. This can break direct writes to the
+	 * framebuffer.  Use this event for cases that cannot be satisfied by FabricBakedModel,
+	 * BlockEntityRenderer or other existing abstraction. If at all possible, use an existing terrain
+	 * RenderLayer instead of outputting to the framebuffer directly with GL calls.
+	 *
+	 * <p>The consumer is responsible for setup and tear down of GL state appropriate for the intended output.
+	 *
+	 * <p>Because solid and cutout quads are depth-tested, order of output does not matter except to improve
+	 * culling performance, which should not be significant after primary terrain rendering. This means
+	 * mods that currently hook calls to individual render layers can simply execute them all at once when
+	 * the event is called.
+	 *
+	 * <p>This event fires before entities and block entities are rendered and may be useful to prepare them.
```

Comment:
Clarify that this event is not only called before E/BEs are rendered, but also before their render submits are created. I also do not understand what is meant by "prepare them"; if this has anything to do with the E/BE render state, then this event should not be used to interact with it.